### ISSUE TITLE
Invalid sub-query SQL generated

### ISSUE THREAD

**Issue by brd-mikek** [linq2db/linq2db#4160]
Please see demo project attached.
We are facing an issue where when we're trying to use sub-query, Linq2Db is creating an invalid SQL query.
Below is the query that gets executed when we run the scenario. 
As you can see, the inner select statement returns 2 values in this scenario which results in a "ORA-00913: too many values" error.

------Linq------

```
var data = (
    from pe in db.PERSON
    select new
    {
        CURRENCYCALCULATEDSYMBOL = (from cc in db.CITY
                                    where cc.CITYCODE == pe.CITYCODE
                                    select cc.CITYNAME).FirstOrDefault(),
    }).Distinct().ToList();
```

------Linq------

------Generated SQL------

```
DECLARE @take Int32
SET     @take = 1

SELECT DISTINCT
        (
                SELECT
                        cc.CITYNAME,
                        1
                FROM
                        CITY cc
                WHERE
                        (cc.CITYCODE = pe.CITYCODE OR cc.CITYCODE IS NULL AND pe.CITYCODE IS NULL)
                FETCH NEXT :take ROWS ONLY
        ),
        t1.CITYNAME,
        t1."is_empty"
FROM
        PERSON pe
                OUTER APPLY (
                        SELECT
                                cc.CITYNAME,
                                1 as "is_empty"
                        FROM
                                CITY cc
                        WHERE
                                (cc.CITYCODE = pe.CITYCODE OR cc.CITYCODE IS NULL AND pe.CITYCODE IS NULL)
                        FETCH NEXT :take ROWS ONLY
                ) t1
```

------Generated SQL------

[LinqToDbIssueDemo.zip](https://github.com/linq2db/linq2db/files/11683423/LinqToDbIssueDemo.zip)

**Comment by brd-mikek** [comment]
Another question. Your old versions used to generate sub-queries and your new versions generate OUTER APPLY clauses. I liked sub-queries better. Is there a way to choose old behavior?

**Comment by MaceWindu** [comment]
Try to add `AsSubquery()` call, maybe it will help

**Comment by brd-mikek** [comment]
Hey guys, any update on this?
We want to upgrade our code from old version (2.9.6) for that we need a couple of bugs resolved - this one and also 4163 and 4167.

**Comment by MaceWindu** [comment]
Will try to check it in next couple of weeks. Don't have much time currently

**Comment by sdanyliv** [comment]
Try the following query as workaround: 
```cs
var data = (
    from pe in db.PERSON
    from cc in db.CITY
      .Where(cc => cc.CITYCODE == pe.CITYCODE)
      .Take(1)
      .DefaultIfEmpty()
    select new
    {
        CURRENCYCALCULATEDSYMBOL = cc.CITYNAME,
    }).Distinct().ToList();
```

**Comment by brd-mikek** [comment]
> Try the following query as workaround:
> 
> ```cs
> var data = (
>     from pe in db.PERSON
>     from cc in db.CITY
>       .Where(cc => cc.CITYCODE == pe.CITYCODE)
>       .Take(1)
>       .DefaultIfEmpty()
>     select new
>     {
>         CURRENCYCALCULATEDSYMBOL = cc.CITYNAME,
>     }).Distinct().ToList();
> ```

Thanks for that.
However problem is not in writing new queries.
We have (big!) codebase that we want to upgrade to new version.
We need to be sure that our existing code still works after upgrade.

**Comment by MaceWindu** [comment]
Change happended in [3.0.0-preview.2}(https://github.com/linq2db/linq2db/milestone/36?closed=1)

Old  sql
```sql
SELECT DISTINCT
        (
                SELECT
                        cc_1.CITYNAME
                FROM
                        CITY cc_1
                WHERE
                        ROWNUM <= 1 AND (cc_1.CITYCODE IS NULL AND pe.CITYCODE IS NULL OR cc_1.CITYCODE = pe.CITYCODE)
        ),
        (
                SELECT
                        cc_2.CITYNAME
                FROM
                        CITY cc_2
                WHERE
                        ROWNUM <= 1 AND (cc_2.CITYCODE IS NULL AND pe.CITYCODE IS NULL OR cc_2.CITYCODE = pe.CITYCODE)
        )
FROM
        PERSON pe
```

**Comment by MaceWindu** [comment]
Probably by this PR https://github.com/linq2db/linq2db/pull/2087

**Comment by MaceWindu** [comment]
Actually both queries (old version and new) look strange. Providers without APPLY support generate more sane SQL:

```sql
SELECT DISTINCT
        (
                SELECT
                        cc.CITYNAME
                FROM
                        CITY cc
                WHERE
                        (cc.CITYCODE = pe.CITYCODE OR cc.CITYCODE IS NULL AND pe.CITYCODE IS NULL) AND
                        ROWNUM <= :take
        )
FROM
        PERSON pe
```

**Comment by brd-mikek** [comment]
Honestly, I'm not sure why you added support for APPLY.
It seems to me that the same can be achieved with simple JOIN.
Better to have one approach than two in my opinion.

Wondering what was your motivation?
Wondering also, are there options to generate JOIN instead of APPLY?

**Comment by MaceWindu** [comment]
While APPLY is not being necessary in this specific case, we still need it for more complex queries.
You can set `PreferApply` option to false to not force APPLY when we can use joins

**Comment by brd-mikek** [comment]
Hey guys!
Any update on this issue by any chance?
We raised this one and a couple others (4163, 4167, 4172) and they stop us from upgrading to latest version.

**Comment by brd-mikek** [comment]
Hi guys!
This bug was open for more than 3 month.
Wondering if this project is still supported?
Is there any way we can contribute?

**Comment by MaceWindu** [comment]
That is fixed now in 6.0 preview 1
